versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/IndexesLogicImpl.java

```diff
@@ -0,0 +1,535 @@
+            "Commit %s references a reference index, which is of unsupported key index type "
+                + indexType);
+    }
+    return referenceIndex;
+  }
+
+  static StoreIndex<CommitOp> deserializeIndex(ByteString serialized) {
+    return deserializeStoreIndex(serialized, COMMIT_OP_SERIALIZER, (k, c) -> c);
+  }
+
+  private StoreIndex<CommitOp> loadIndexSegment(
+      @Nonnull @jakarta.annotation.Nonnull ObjId indexId) {
+    IndexObj index;
+    try {
+      index = persist.fetchTypedObj(indexId, ObjType.INDEX, IndexObj.class);
+    } catch (ObjNotFoundException e) {
+      throw new IllegalStateException(
+          format("Commit %s references a reference index, which does not exist", indexId));
+    }
+    return deserializeIndex(index.index()).setObjId(indexId);
+  }
+
+  private StoreIndex<CommitOp>[] loadIndexSegments(
+      @Nonnull @jakarta.annotation.Nonnull ObjId[] indexes) {
+    try {
+      Obj[] objs = persist.fetchObjs(indexes);
+      @SuppressWarnings("unchecked")
+      StoreIndex<CommitOp>[] r = new StoreIndex[indexes.length];
+      for (int i = 0; i < objs.length; i++) {
+        Obj obj = objs[i];
+        if (obj != null) {
+          IndexObj index = (IndexObj) obj;
+          r[i] = deserializeIndex(index.index()).setObjId(indexes[i]);
+        }
+      }
+      return r;
+    } catch (ObjNotFoundException e) {
+      throw new IllegalStateException(format("Reference index segments %s not found", e.objIds()));
+    }
+  }
+
+  private StoreIndex<CommitOp> referenceIndexFromStripes(
+      List<IndexStripe> indexStripes, ObjId commitId) {
+    List<StoreIndex<CommitOp>> stripes = new ArrayList<>(indexStripes.size());
+    List<StoreKey> firstLastKeys = new ArrayList<>(indexStripes.size() * 2);
+
+    @SuppressWarnings("unchecked")
+    StoreIndex<CommitOp>[] loaded = new StoreIndex[indexStripes.size()];
+
+    for (int i = 0; i < indexStripes.size(); i++) {
+      IndexStripe s = indexStripes.get(i);
+      int idx = i;
+      stripes.add(
+          lazyStoreIndex(
+                  () -> {
+                    StoreIndex<CommitOp> l = loaded[idx];
+                    if (l == null) {
+                      LOGGER.debug(
+                          "Individual fetch of stripe #{} of {} stripes for commit {}",
+                          idx,
+                          loaded.length,
+                          commitId);
+                      l = loadIndexSegment(s.segment());
+                      loaded[idx] = l;
+                    }
+                    return l;
+                  },
+                  s.firstKey(),
+                  s.lastKey())
+              .setObjId(s.segment()));
+      firstLastKeys.add(s.firstKey());
+      firstLastKeys.add(s.lastKey());
+    }
+    if (stripes.size() == 1) {
+      return stripes.get(0);
+    }
+
+    IndexLoader<CommitOp> indexLoader =
+        indexesToLoad -> {
+          checkArgument(indexesToLoad.length == loaded.length);
```

Comment:
This (bulk) loader and the loader inside the lazy stripes essentially load the same data just on different call paths but their "scope of responsibility" is essentially the same. WDYT about manging stripe their loading via one  `IndexLoader`? I suppose we could have the lazy index load its data via the same loader interface.

Reply:
(Not sure I completely understood your suggesting)

Usually, the call paths load the required index stripes eagerly, based on the requested store-keys, so that lazy loading should practically never happen.

IIRC there were some rare scenarios, in which it could be theoretically possible (don't really recall which TBH).

Duplicate loads of the same obj-IDs are mitigated, if the cache layer's present.

Reply:
My suggestion is to use the same `IndexLoader` impl both for the bulk load case and for the single lazy load case.

Reply:
the "load N" and "load single" implementations are quite different - unifying those is likely a bit of an over-generalization